feat(sync): enable thread-safe and de-duplicated syncer registration

powerslider opened the pull request:
## Why this should be merged

Adds crucial validations that disallow misuse of the `SyncerRegistry`.

## How this works

- Replace slice-based storage with `sync.Map` for concurrent access.
- Use `LoadOrStore` for atomic duplicate prevention during registration.
- Add validation that you can call `RunSyncerTasks` only once.
- Update tests to work with a concurrent model and verify thread safety.
- Add validation for the presence of a client summary.
- Add frozen `atomic.Value` to hold immutable `[]SyncerTask` snapshot.
- On first `RunSyncerTasks`, collect from `sync.Map` once and store in `frozen`.
  Iterate `frozen` slice instead of ranging `sync.Map` to avoid per-call copying.
- This is mostly prep work for when introducing `UpdateSyncTarget` in
  the `Syncer` API to support Firewood's changesets.

## How this was tested
Test cases were added in `sync/registry_test.go`.

## Need to be documented?
no

## Need to update RELEASES.md?
no 

resolves ava-labs/avalanchego#4601

Signed-off-by: Tsvetan Dimitrov (tsvetan.dimitrov@avalabs.org)

powerslider committed:
feat(sync): enable thread-safe syncer registration

- Replace slice-based storage with `sync.Map` for concurrent access.
- Use `LoadOrStore` for atomic duplicate prevention during registration.
- Add validation that you can call `RunSyncerTasks` only once.
- Update tests to work with concurrent model and verify thread safety.
- Add validation for the presence of a client summary.

resolves #1098

Signed-off-by: Tsvetan Dimitrov (tsvetan.dimitrov@avalabs.org)

alarso16 commented on sync/vm/registry.go:
```diff
@@ -19,37 +31,70 @@ type SyncerTask struct {
-	if r.registeredNames[name] {
-		return fmt.Errorf("syncer with name '%s' is already registered", name)
+	// If the syncer registry has already been started, return an error.
+	if r.started.Load() {
```
I've thought about this more: I still don't think it's necessary for all of this code to be thread-safe, but parts of it do during dynamic state sync, so it does seem easier to guarantee correctness is if it's all thread-safe

alarso16 commented on sync/vm/registry.go:
```diff
@@ -19,37 +31,70 @@ type SyncerTask struct {
+
+	// Collect all syncers from the map.
+	var syncers []SyncerTask
+	r.syncers.Range(func(key, value any) bool {
```
I don't think this is a good way to store the `syncers`. Eventually, we're also going to have another function like this:
```go
func (r *SyncerRegistry) UpdateSyncTarget(ctx context.Context, block /* unsure of type */) error {
    for name, task := syncTasks {
        if block.Height() % task.interval == 0 {
            log.Info("Updating sync target", "syncer", name, "height", block.Height())
            if err := task.syncer.UpdateSyncTarget(ctx, block) {
                // handle err somehow
            }
        }
    }
    return nil
}
```
I don't think it's a good idea to have to re-construct the array every time we call this (approx. every 2 seconds for a couple hours)

powerslider replied:
Changes solving this problem you can check [here](c632a4dc9dfa3d7d6e76430974a3c35cd98e37b5)

alarso16 commented on sync/vm/registry.go:
```diff
@@ -5,12 +5,24 @@ package vm
 )
 
+var (
+	errClientCannotProvideSummary = errors.New("client cannot provide a summary")
```
```suggestion
	errClientCannotProvideSummary = errors.New("client must provide a summary")
```
I think this will ultimately be whatever the most recent `target` is, but a sync summary is ok for now

powerslider commented on sync/vm/registry_test.go:
```diff
@@ -6,25 +6,33 @@ package vm
-	startCallCount int
-	waitCallCount  int
-	started        bool // Track if already started
+	startCalled    atomic.Bool
```
We are not testing syncers here, this is just to test the behaviour of the registry. Making the `mockSyncer` thread safe is just because of running the tests in parallel. This has no connection to the syncers themselves.

powerslider committed:
perf(sync): freeze registered syncers on start for lock-free iteration

- Add frozen atomic.Value to hold immutable []SyncerTask snapshot.
- Initialize frozen with empty slice in NewSyncerRegistry.
- On first RunSyncerTasks, collect from sync.Map once and store in frozen.
  Iterate frozen slice instead of ranging sync.Map to avoid per-call copying.
- This is mostly prep work for when introducing `UpdateSyncTarget` in
  the `Syncer` API to support Firewood's changesets.

ceyonur reviewed (commented):
Could you provide some more context why we need these in the PR description?

powerslider committed:
fix: fix flaky test

powerslider committed:
Merge branch 'master' into powerslider/1098-thread-safe-syncer-registration

powerslider committed:
fix: fix flaky blocksync test

powerslider committed:
Merge branch 'master' into powerslider/1098-thread-safe-syncer-registration

powerslider commented on sync/vm/registry.go:
```diff
@@ -19,37 +31,82 @@ type SyncerTask struct {
+	// Use reflection to check for nil because in Go, a nil concrete type is not equal to a nil interface [synccommon.Syncer].
+	// When a nil concrete type is assigned to an interface, the interface contains type information even though the value is nil.
+	// reflect.ValueOf(syncer).IsNil() properly detects nil concrete types.
+	if syncer == nil || reflect.ValueOf(syncer).IsNil() {
```
If I remove the `reflect` check, then I have to remove the entire check, because it will work for only some cases.

powerslider replied:
I don't mind to remove it, but personally I would be more defensive especially in the registry, because we start counting way too much that externally the caller will validate everything properly for a component that should act as a black box to its consumer.

powerslider committed:
fix: remove empty syncer name check

powerslider committed:
Merge branch 'master' into powerslider/1098-thread-safe-syncer-registration
